platform/jvm/rules.bzl

```diff
@@ -0,0 +1,391 @@
+        ),
+        "main_class": attr.string(
+            mandatory = True,
+            doc = "The main class name, which must be fully qualified (using the full package address)",
+        ),
+        "os": attr.string(
+            mandatory = True,
+            doc = "The host OS",
+        ),
+        "verbose": attr.bool(
+            mandatory = True,
+            doc = "Enable verbose output for debugging",
+        ),
+        "log_sensitive_data": attr.bool(
+            mandatory = True,
+            doc = "If enabled, sensitive data including passwords will be included in the output logs",
+        ),
+        "create_shortcut": attr.bool(
+            mandatory = True,
+            doc = "Whether to create a shortcut for the application (on systems that support desktop shortcuts)",
+        ),
+        "linux_app_category": attr.string(
+            doc = "'Section' value of the DEB control file on Debian-based Linux systems",
+        ),
+        "linux_menu_group": attr.string(
+            doc = "Menu group this application is placed in on Linux, defining the categories under which the application will be classified",
+        ),
+        "mac_app_id": attr.string(
+            doc = "On Mac, the primary bundle ID that uniquely identifies the application, which is submitted to Apple's notarization service to verify the app",
+        ),
+        "mac_entitlements": attr.label(
+            allow_single_file = True,
+            doc = "The Mac entitlements.mac.plist file",
+        ),
+        "mac_code_signing_cert": attr.label(
+            allow_single_file = True,
+            doc = "The Mac code signing certificate",
+        ),
+        "mac_deep_sign_jars_regex": attr.string(
+            doc = "On Mac, JARs in the Java deps whose names match this regex will be repackaged with their native libraries signed",
+        ),
+        "windows_menu_group": attr.string(
+            doc = "Start Menu group this application is placed in on Windows. If unset, the application will not be placed in the Start Menu",
+        ),
+        "windows_wix_toolset": attr.label(
+            allow_single_file = True,
+            doc = "Archive containing the Windows WiX toolset",
+        ),
+        "_assemble_jvm_platform_bin": attr.label(
+            default = "@vaticle_bazel_distribution//platform/jvm:assembler-bin",
+            executable = True,
+            cfg = "host",
+        ),
+    },
+    outputs = {
+        "distribution_file": "%{name}.zip"
+    },
+    implementation = _assemble_zip_to_jvm_platform_impl,
+    doc = "A JVM application image",
+)
+
+
+# TODO: upgrade all to JDK17
+def native_jdk16():
+    return select({
+        "@vaticle_dependencies//util/platform:is_mac": "@jdk16_mac//file",
+        "@vaticle_dependencies//util/platform:is_linux": "@jdk16_linux//file",
+        "@vaticle_dependencies//util/platform:is_windows": "@jdk16_windows//file",
+    })
+
```

Comment:
The `assemble_jvm_platform` macro itself. This is what applications such as Studio actually use, from this file.

Reply:
NOTE: We should document this somewhere:
> if the user gets "The configured main jar does not exist xxx.jar in the input directory",
they should ensure java_deps_root is set the same as in the java_deps, and that the main JAR name is set
to the Maven name (x-y-z-0.0.0.jar where x-y-z is the main class's package address in kebab-case)